File: Practice/Dynamic_Programming/q31/q31.cpp
#include <iostream>
#include <algorithm>
#include <cstring>

using namespace std;

int board[22][20];
int dp[22][20];
int t, n, m;

int main(void){
    cin >> t;
    while(t-- > 0){
        cin >> n >> m;
        int x, y;
        for(int i = 1; i <= n; i++){
            for(int j = 0; j < m; j++){
                cin >> board[i][j];                
                if(j == 0){
                    dp[i][j] = board[i][j];
                }
            }
        }
        
        for(int col = 1; col < m; col++){
            for(int row = 1; row < n+1; row++){
                dp[row][col] =
                 max(dp[row-1][col],
                  max(board[row][col]+dp[row-1][col-1],
                   max(board[row][col]+dp[row][col-1], board[row][col]+dp[row+1][col-1])));
            }
        }
        cout << dp[n][m-1] << "\n";
        memset(dp, 0, sizeof(dp));
    }
}

File: Practice/Dynamic_Programming/q31/q31.py
t = int(input())
while t > 0:
    n, m = map(int, input().split())
    gold = list(map(int, input().split()))
    dp = []
    index = 0
    for i in range(n+2):
        if i == 0 or i == n+1:
            dp.append([0]*m)
            continue
        dp.append(gold[index:index+m])
        index += m
    for col in range(1, m):
        for row in range(1, n+1):
            dp[row][col] = max(dp[row-1][col], dp[row-1][col-1]+dp[row][col], dp[row][col-1]+dp[row][col], dp[row+1][col-1]+dp[row][col])
    t -= 1
    print(dp[n][m-1])